src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenMscorlib.cs

```diff
@@ -952,5 +951,228 @@ .locals init (System.IntPtr V_0)
 }
 ");
         }
+
+        [Fact]
+        public void FixedHelpersOnStringAndArray()
+        {
+            var text =
+@"
+namespace System
+{
+    public struct Void { }
+    public class ValueType { }
+    public struct Int32 { }
+    public struct Char { }
+    public struct Boolean { }
+
+    public class String 
+    { 
+        public ref char GetPinnableReference() => throw null;
+    }
+
+    public class Object 
+    { 
+    }
+
+    public class Array 
+    { 
+        public ref char GetPinnableReference() => throw null;
+    }
+}
+  
+unsafe internal class program
+{
+    public static void Main()
+    {
+        fixed (char* p = ""A"")
+        {
+            *p = default;
+        }
+
+        fixed (char* p = new char[1])
+        {
+            *p = default;
+        }
+    }
+}
+";
+            var comp = CreateEmptyCompilation(text, options: TestOptions.UnsafeReleaseDll);
+
+            comp.VerifyDiagnostics();
+
+
+            CompileAndVerify(comp, verify: Verification.Skipped).
+                VerifyIL("program.Main()", @"
```

Comment:
Could you split Main into two methods (one for string case, and the other for array case) so that it's clear from IL which method is called and which is not?